src/transformers/models/bark/modeling_bark.py

```diff
@@ -0,0 +1,1372 @@
+        super().__init__(*inputs, **kwargs)
+
+
+class BarkModulePreTrainedModel(PreTrainedModel):
+    """
+    An abstract class to handle weights initialization and a simple interface for downloading and loading pretrained
+    models.
+    """
+
+    config_class = BarkModuleConfig
+    # supports_gradient_checkpointing = True
+    _no_split_modules = ["BarkBlock"]  # TODO: what to do with this?
+
+    def _init_weights(self, module):
+        """Initialize the weights."""
+        if isinstance(module, (nn.Linear,)):
+            # Slightly different from the TF version which uses truncated_normal for initialization
+            # cf https://github.com/pytorch/pytorch/pull/5617
+            module.weight.data.normal_(mean=0.0, std=self.config.initializer_range)
+            if module.bias is not None:
+                module.bias.data.zero_()
+        elif isinstance(module, nn.Embedding):
+            module.weight.data.normal_(mean=0.0, std=self.config.initializer_range)
+            if module.padding_idx is not None:
+                module.weight.data[module.padding_idx].zero_()
+        elif isinstance(module, nn.LayerNorm):
+            module.bias.data.zero_()
+            module.weight.data.fill_(1.0)
+
+    def _set_gradient_checkpointing(self, module, value=False):
+        if isinstance(module, BarkCausalModule) or isinstance(module, BarkFineAcousticsModule):
+            module.gradient_checkpointing = value
+
+
+# GPT2-like autoregressive model
+@add_start_docstrings(
+    "Bark sub-module at the core of the semantic and the coarse acoustics sub-models. It is a GPT-2 like autoregressive model with a language modeling head on top.",
+    BARK_MODULE_START_DOCSTRING,
+)
+class BarkCausalModule(BarkModulePreTrainedModel):
+    # TODO: add code sample when checkpoint is added
+    #
+    # @add_code_sample_docstrings(
+    #    checkpoint=_CHECKPOINT_FOR_DOC,
+    #    output_type=BaseModelOutputWithPast,
+    #    config_class=_CONFIG_FOR_DOC,
+    # )
+
+    def __init__(self, config):
+        super().__init__(config)
+        self.config = config
+
+        self._initialize_modules(config)
+
+        self.gradient_checkpointing = False
+
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    def _initialize_modules(self, config):
+        # initialize as an autoregressive GPT-like model
+        self.transformer = nn.ModuleDict(
+            {
+                "wte": nn.Embedding(config.input_vocab_size, config.hidden_size),
+                "wpe": nn.Embedding(config.block_size, config.hidden_size),
+                "drop": nn.Dropout(config.dropout),
+                "h": nn.ModuleList([BarkBlock(config, idx, is_causal=True) for idx in range(config.num_layers)]),
+                "ln_f": LayerNorm(config.hidden_size, bias=config.bias),
+            }
+        )
+        self.lm_head = nn.Linear(config.hidden_size, config.output_vocab_size, bias=False)
```

Comment:
Would be in favour of defining this as a base model in this class, then adding an LM head in each new class that comes after it (see GPT Neo design for example)